runtime/tracing/tracing-kotlin-logging/common/src/aws/smithy/kotlin/runtime/tracing/KotlinLoggingTraceProbe.kt

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.tracing
+
+import mu.KLogger
+import mu.KotlinLogging
+
+/**
+ * A [TraceProbe] that logs events to the [kotlin-logging](https://github.com/MicroUtils/kotlin-logging) library.
+ * kotlin-logging requires configuration (e.g., of Slf4j on JVM) before events sent to the probe will appear in logs.
+ */
+public class KotlinLoggingTraceProbe : TraceProbe {
+    private fun EventLevel.loggerMethod(): (KLogger, () -> Any?) -> Unit = when (this) {
+        EventLevel.Fatal,
+        EventLevel.Error, -> KLogger::error
+        EventLevel.Warning -> KLogger::warn
+        EventLevel.Info -> KLogger::info
+        EventLevel.Debug -> KLogger::debug
+        EventLevel.Trace -> KLogger::trace
+    }
+
+    private fun log(spanId: String, event: TraceEvent) {
+        val logger = KotlinLogging.logger(event.sourceComponent)
+        val method = event.level.loggerMethod()
+        method(logger) {
+            val msg = (event.data as TraceEventData.Message).content()
+            "$spanId: $msg"
+        }
+    }
+
+    override fun postEvents(span: TraceSpan, events: Iterable<TraceEvent>) {
+        val spanId = span.hierarchicalId
+        events.filter { it.data is TraceEventData.Message }.forEach { log(spanId, it) }
```

Comment:
filter creates a new `List` instance IIRC

Given how often this will be called we may want to just iterate over the events and do the check in the loop:

```kotlin
events.forEach { if (it.data is TraceEventData.Message) log(spanId, it) }
```